crates/amaru-consensus/src/consensus/effects/mempool_effects.rs

```diff
@@ -0,0 +1,393 @@
+// limitations under the License.
+
+use amaru_kernel::Tx;
+use amaru_ouroboros_traits::{
+    CanValidateTransactions, MempoolSeqNo, TransactionValidationError, TxId, TxOrigin,
+    TxRejectReason, TxSubmissionMempool,
+};
+use pure_stage::{
+    BoxFuture, Effects, ExternalEffect, ExternalEffectAPI, ExternalEffectSync, Resources, SendData,
+};
+use serde::{Deserialize, Serialize};
+use std::fmt::Debug;
+use std::pin::Pin;
+use std::sync::Arc;
+
+/// Implementation of Mempool using pure_stage::Effects.
+#[derive(Clone)]
+pub struct MemoryPool<T> {
+    effects: Effects<T>,
+}
+
+impl<T> MemoryPool<T> {
+    pub fn new(effects: Effects<T>) -> MemoryPool<T> {
+        MemoryPool { effects }
+    }
+
+    /// This function runs an external effect synchronously.
+    pub fn external_sync<E: ExternalEffectSync + serde::Serialize + 'static>(
+        &self,
+        effect: E,
+    ) -> E::Response
+    where
+        T: SendData + Sync,
+    {
+        self.effects.external_sync(effect)
+    }
+}
+
+impl<T: SendData + Sync> CanValidateTransactions<Tx> for MemoryPool<T> {
+    fn validate_transaction(&self, tx: &Tx) -> Result<(), TransactionValidationError> {
+        self.effects.external_sync(ValidateTransaction(tx.clone()))
+    }
+}
+
+impl<T: SendData + Sync> TxSubmissionMempool<Tx> for MemoryPool<T> {
+    fn insert(&self, tx: Tx, tx_origin: TxOrigin) -> Result<(TxId, MempoolSeqNo), TxRejectReason> {
+        self.external_sync(Insert::new(tx, tx_origin))
+    }
+
+    fn get_tx(&self, tx_id: &TxId) -> Option<Arc<Tx>> {
+        self.external_sync(GetTx::new(tx_id.clone()))
+    }
+
+    fn tx_ids_since(&self, from_seq: MempoolSeqNo, limit: u16) -> Vec<(TxId, u32, MempoolSeqNo)> {
+        self.external_sync(TxIdsSince::new(from_seq, limit))
+    }
+
+    fn wait_for_at_least(
+        &self,
+        seq_no: MempoolSeqNo,
+    ) -> Pin<Box<dyn Future<Output = bool> + Send + '_>> {
+        self.effects.external(WaitForAtLeast::new(seq_no))
+    }
+
+    fn get_txs_for_ids(&self, ids: &[TxId]) -> Vec<Arc<Tx>> {
+        self.external_sync(GetTxsForIds::new(ids))
+    }
+
+    fn last_seq_no(&self) -> MempoolSeqNo {
+        self.external_sync(LastSeqNo)
+    }
+}
+
+// EXTERNAL EFFECTS DEFINITIONS
+
+pub type ResourceMempool = Arc<dyn TxSubmissionMempool<Tx>>;
+
+#[derive(Debug, PartialEq, Eq, Serialize, Deserialize)]
+struct Insert {
+    tx: Tx,
+    tx_origin: TxOrigin,
+}
+
+impl Insert {
+    pub fn new(tx: Tx, tx_origin: TxOrigin) -> Self {
+        Self { tx, tx_origin }
+    }
+}
+
+impl ExternalEffect for Insert {
+    #[expect(clippy::expect_used)]
+    fn run(self: Box<Self>, resources: Resources) -> BoxFuture<'static, Box<dyn SendData>> {
+        Self::wrap_sync({
+            let mempool = resources
+                .get::<ResourceMempool>()
+                .expect("ResourceMempool requires a mempool")
+                .clone();
+            mempool.insert(self.tx, self.tx_origin)
```

Comment:
I see, so you’re using `trait TxSubmissionMempool<Tx>` for two very different purposes. My gut feeling is that this will lead to confusion, the API trait and the resource implementation should be different types. And the resource implementation doesn’t need to go through a trait because effects can already be replaced in pure-stage.